File: pacific-adapter/adapter/src/main/java/com/pacific/adapter/ViewHolder.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pacific.adapter;

import android.support.annotation.IdRes;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class ViewHolder extends RecyclerView.ViewHolder implements ListenerAttach {

    /**
     * binding
     */
    private Object binding;
    /**
     * adapter position
     */
    private int position = -1;

    /**
     * adapter size
     */
    private int size = 0;

    /**
     * item
     */
    private Item item;

    /**
     * ListenerAttach
     */
    private final ListenerAttachImpl listenerAttach;

    public ViewHolder(View itemView, ListenerProvider provider) {
        super(itemView);
        this.listenerAttach = new ListenerAttachImpl(provider, this);
        try {
            Class<?> clazz = Class.forName("android.databinding.DataBindingUtil");
            Method method = clazz.getMethod("bind", View.class);
            binding = method.invoke(null, itemView);
        } catch (ClassNotFoundException e) {
            binding = new DefaultBinding(itemView);
        } catch (NoClassDefFoundError error) {
            binding = new DefaultBinding(itemView);
        } catch (NoSuchMethodException e) {
            binding = new DefaultBinding(itemView);
        } catch (IllegalAccessException e) {
            binding = new DefaultBinding(itemView);
        } catch (InvocationTargetException e) {
            binding = new DefaultBinding(itemView);
        } catch (IllegalArgumentException e) {
            binding = new DefaultBinding(itemView);
        }
    }

    /**
     * get adapter size
     *
     * @return
     */
    public int getSize() {
        return size;
    }

    /**
     * get item
     *
     * @return
     */
    public <V extends Item> V getItem() {
        return (V) item;
    }

    /**
     * is first item
     *
     * @return
     */
    public boolean isFirstItem() {
        if (position == 0) return true;
        return false;
    }

    /**
     * is last item
     *
     * @return
     */
    public boolean isLastItem() {
        if (position == size - 1) return true;
        return false;
    }

    /**
     * get adapter position , just used in ListView,GridView,Spinner,ViewPager
     *
     * @return
     */
    public int getCurrentPosition() {
        return position;
    }

    void setCurrentPosition(int position) {
        this.position = position;
    }

    void setSize(int size) {
        this.size = size;
    }

    void setCurrentItem(Item item) {
        this.item = item;
    }

    /**
     * get binding,if you are using DataBinding , it will return ViewDataBinding,
     * or it will return {@link DefaultBinding}
     *
     * @param <T>
     * @return
     */
    public <T extends Object> T binding() {
        return (T) this.binding;
    }

    /**
     * attach OnClickListener for view
     *
     * @param viewId view id
     */
    @Override
    public void attachOnClickListener(@IdRes int viewId) {
        listenerAttach.attachOnClickListener(viewId);
    }

    /**
     * attach OnTouchListener for view
     *
     * @param viewId view id
     */
    @Override
    public void attachOnTouchListener(@IdRes int viewId) {
        listenerAttach.attachOnTouchListener(viewId);
    }

    /**
     * attach OnLongClickListener for view
     *
     * @param viewId view id
     */
    @Override
    public void attachOnLongClickListener(@IdRes int viewId) {
        listenerAttach.attachOnLongClickListener(viewId);
    }

    /**
     * attach CompoundButton.OnCheckedChangeListener for CompoundButton
     *
     * @param viewId CompoundButton view id
     */
    @Override
    public void attachOnCheckedChangeListener(@IdRes int viewId) {
        listenerAttach.attachOnCheckedChangeListener(viewId);
    }

    @Override
    public void attachImageLoader(@IdRes int viewId) {
        listenerAttach.attachImageLoader(viewId);
    }
}

File: pacific-adapter/adapter/src/main/java/com/pacific/adapter/ViewHolder.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.pacific.adapter

import android.app.Activity
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IdRes
import androidx.recyclerview.widget.RecyclerView

class ViewHolder(
    itemView: View,
    provider: ListenerProvider
) : RecyclerView.ViewHolder(itemView), ListenerAttach {

    private val listenerAttach: ListenerAttachImpl = ListenerAttachImpl(this, provider)
    private val binding: Any by lazy {
        try {
            val clazz: Class<*> = try {
                Class.forName("androidx.databinding.DataBindingUtil")
            } catch (ignored: Exception) {
                Class.forName("android.databinding.DataBindingUtil")
            }
            clazz.getMethod("bind", View::class.java).invoke(null, itemView)
        } catch (ignored: Exception) {
            DefaultBinding(itemView)
        }
    }

    var size: Int = -1
        internal set

    var itemPosition: Int = -1
        internal set

    lateinit var item: RecyclerItem<*>
        internal set

    val isFirstItem: Boolean
        get() = itemPosition == 0

    val isLastItem: Boolean
        get() = itemPosition == size - 1

    override fun attachOnClickListener(@IdRes viewId: Int) {
        listenerAttach.attachOnClickListener(viewId)
    }

    override fun attachOnTouchListener(@IdRes viewId: Int) {
        listenerAttach.attachOnTouchListener(viewId)
    }

    override fun attachOnLongClickListener(@IdRes viewId: Int) {
        listenerAttach.attachOnLongClickListener(viewId)
    }

    override fun attachOnCheckedChangeListener(@IdRes viewId: Int) {
        listenerAttach.attachOnCheckedChangeListener(viewId)
    }

    override fun attachImageLoader(@IdRes viewId: Int) {
        listenerAttach.attachImageLoader(viewId)
    }

    override fun attachTextChangedListener(@IdRes viewId: Int) {
        listenerAttach.attachTextChangedListener(viewId)
    }

    override fun detachTextChangedListener(@IdRes viewId: Int) {
        listenerAttach.detachTextChangedListener(viewId)
    }

    @Suppress("UNCHECKED_CAST")
    fun <T : RecyclerItem<*>> item(): T = item as T

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> binding(): T = binding as T

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> adapter(): T = listenerAttach.provider as T

    @Suppress("UNCHECKED_CAST")
    fun <T : ViewGroup> parentViewGroup(): T = itemView.parent as T

    @Suppress("UNCHECKED_CAST")
    fun <T : Activity> activity(): T = itemView.context as T
}